ssh-mitm: 3.0.2 → 4.1.1

AkechiShiro opened the pull request:
* Hash wasn't updated, updating the hash to the correct value seems to make the build fail with a very weird error. (EDIT : Hash was finally updated)
* Paramiko override necessary, (EDIT as last release needs Paramiko <3.4/>=3.3.0).

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

@fabaff 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AkechiShiro committed:
ssh-mitm: 3.0.2 → 4.1.1

* Hash update
* Add tkinter dependency in propagatedBuildInputs and remove setuptools from
  propagatedBuildInputs, add to natievBuildInputs
* Format using nixfmt so nix is parsable by GitHub CI/CD ??
* Fix python3Packages to python3.pkgs
* Fix for ssh-mitm askpass
* Workaround for darwin users
* Import stdenv missing
* Fix stdenv.isDarwin
* Apply fix from fabaff's review (man still missing)
* Attempt to add man pages
* Override paramiko, patch test_config for paramiko 3.11 (+), due to deprecated
  pytest setup method

AkechiShiro commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -3,36 +3,22 @@
     rev = "refs/tags/${version}";
     hash = "sha256-koV7g6ZmrrXk60rrDP8BwrDZk3shiyJigQgNcb4BASE=";
+    # Correct hash does not seems to build ??
+    #hash = "sha256-Uf1B7oEZyNWj4TjrLvEfFdxsvsGeMLXFsSdxGLUV4ZU=";
```
This is the proper hash of the package but I'm running into a failure : 
```
❯ nix-build -A ssh-mitm
this derivation will be built:
  /nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv
building '/nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Running phase: unpackPhase
unpacking source archive /nix/store/rksl0phb64vhv1l4yq0iw2shb8vwgpl1-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/sshmitm/workarounds/transport.py
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
no configure script, doing nothing
Running phase: buildPhase
Executing setuptoolsBuildPhase
Finished executing setuptoolsBuildPhase
Running phase: installPhase
Executing pypaInstallPhase
/nix/store/dd3gsp3cvl6l641h4vcs86l9frznxsjx-pypa-install-hook/nix-support/setup-hook: line 8: pushd: dist: No such file or directory
error: builder for '/nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv' failed with exit code 1;
       last 10 log lines:
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > Executing setuptoolsBuildPhase
       > Finished executing setuptoolsBuildPhase
       > Running phase: installPhase
       > Executing pypaInstallPhase
       > /nix/store/dd3gsp3cvl6l641h4vcs86l9frznxsjx-pypa-install-hook/nix-support/setup-hook: line 8: pushd: dist: No such file or directory
       For full logs, run 'nix log /nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv'.
```

When using the original hash, the build is a success.

fabaff replied:
The build system is now `hatchling` and no `setuptools`.

```
hash = "sha256-Uf1B7oEZyNWj4TjrLvEfFdxsvsGeMLXFsSdxGLUV4ZU=";
```

should work then.

AkechiShiro commented:
Could you please review @D3vil0p3r ?

Binary functionality was tested using a local server.

Aleksanaa commented:
> Hash wasn't updated, updating the hash to the correct value seems to make the build fail with a very weird error.

Wait what? By not updating the hash you are not updating the source!

D3vil0p3r commented:
When running `ssh-mitm-askpass` I get:
```
ERROR:root:tkinter not installed!
```
probably some dependency is missing?

D3vil0p3r commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -3,36 +3,22 @@
+    # Correct hash does not seems to build ??
+    #hash = "sha256-Uf1B7oEZyNWj4TjrLvEfFdxsvsGeMLXFsSdxGLUV4ZU=";
   };
 
```
```suggestion

nativeBuildInputs = [ python3Packages.setuptools];

```

D3vil0p3r reviewed (changes requested):
You need also to remove `setuptools` from `propagatedBuldInnputs` and add `tkinter`. I don't know why GitHub does not allow me to add suggestions on "greyed out" code.

AkechiShiro commented:
> > Hash wasn't updated, updating the hash to the correct value seems to make the build fail with a very weird error.
> 
> Wait what? By not updating the hash you are not updating the source!

Yes but updating the hash breaks the build, and I have no clue why.

AkechiShiro commented:
Any more stuff I should fix ? @D3vil0p3r ?

@Aleksanaa : the issue is up there : 
> This is the proper hash of the package but I'm running into a failure :
> 
> ```nix
> ❯ nix-build -A ssh-mitm
> this derivation will be built:
>   /nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv
> building '/nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv'...
> Sourcing python-remove-tests-dir-hook
> Sourcing python-catch-conflicts-hook.sh
> Sourcing python-remove-bin-bytecode-hook.sh
> Sourcing setuptools-build-hook
> Using setuptoolsBuildPhase
> Sourcing pypa-install-hook
> Using pypaInstallPhase
> Sourcing python-imports-check-hook.sh
> Using pythonImportsCheckPhase
> Sourcing python-namespaces-hook
> Sourcing python-catch-conflicts-hook.sh
> Running phase: unpackPhase
> unpacking source archive /nix/store/rksl0phb64vhv1l4yq0iw2shb8vwgpl1-source
> source root is source
> setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/sshmitm/workarounds/transport.py
> Running phase: patchPhase
> Running phase: updateAutotoolsGnuConfigScriptsPhase
> Running phase: configurePhase
> no configure script, doing nothing
> Running phase: buildPhase
> Executing setuptoolsBuildPhase
> Finished executing setuptoolsBuildPhase
> Running phase: installPhase
> Executing pypaInstallPhase
> /nix/store/dd3gsp3cvl6l641h4vcs86l9frznxsjx-pypa-install-hook/nix-support/setup-hook: line 8: pushd: dist: No such file or directory
> error: builder for '/nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv' failed with exit code 1;
>        last 10 log lines:
>        > Running phase: patchPhase
>        > Running phase: updateAutotoolsGnuConfigScriptsPhase
>        > Running phase: configurePhase
>        > no configure script, doing nothing
>        > Running phase: buildPhase
>        > Executing setuptoolsBuildPhase
>        > Finished executing setuptoolsBuildPhase
>        > Running phase: installPhase
>        > Executing pypaInstallPhase
>        > /nix/store/dd3gsp3cvl6l641h4vcs86l9frznxsjx-pypa-install-hook/nix-support/setup-hook: line 8: pushd: dist: No such file or directory
>        For full logs, run 'nix log /nix/store/2yl03yllx3z5w9bd8bfrp6q0wq9mrfwa-ssh-mitm-4.1.1.drv'.
> ```
> 
> When using the original hash, the build is a success.


If you have any idea how I can fix this, I'm okay to give it a try but I haven't been able to understand the error, I've done prior research in older Nixpkgs issues and still don't understand what is going wrong here.

AkechiShiro commented:
Still ran into this error : 
```python
Traceback (most recent call last):
  File "/nix/store/m936y5z7zfkmyj2y65v2bbh1fjhrmgf0-ssh-mitm-4.1.1/bin/.ssh-mitm-wrapped", line 6, in <module>
    from sshmitm.cli import main
  File "/nix/store/m936y5z7zfkmyj2y65v2bbh1fjhrmgf0-ssh-mitm-4.1.1/lib/python3.11/site-packages/sshmitm/cli.py", line 46, in <module>
    from sshmitm.moduleparser import ModuleParser
  File "/nix/store/m936y5z7zfkmyj2y65v2bbh1fjhrmgf0-ssh-mitm-4.1.1/lib/python3.11/site-packages/sshmitm/moduleparser.py", line 39, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'
```

AkechiShiro commented:
@D3vil0p3r The fix was to put `setuptools` as `propagatedBuildInputs`, if not `pkg_resources` is not available for `ssh-mitm-askpass`

There might be a better way to fix it tho, I'd need to dig into Python attributes.

AkechiShiro commented:
Nixpkgs-review output : 
```nix
Automatic merge went well; stopped before committing as requested
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home/kenshin/.cache/nixpkgs-review/pr-308157-7/nixpkgs nixpkgs-overlays=/tmp/tmpq91akokq -qaP --xml --out-path --show-trace --no-allow-import-from-derivation --meta
9 packages updated:
exaile gramps python3.11-bsddb3 python3.11-pymilter python3.11-scrapy-deltafetch python3.12-bsddb3 python3.12-pymilter python3.12-scrapy-deltafetch spf-engine

$ nix build --nix-path nixpkgs=/home/kenshin/.cache/nixpkgs-review/pr-308157-7/nixpkgs nixpkgs-overlays=/tmp/tmpq91akokq --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/kenshin/.cache/nixpkgs-review/pr-308157-7/build.nix
error: builder for '/nix/store/4b0yac479mdwca75ggwa4d276r1xhzwc-python3.12-pymilter-1.0.5.drv' failed with exit code 1;
       last 10 log lines:
       >     self.assertEquals(cache['temp@bar.com'],'testing')
       >     ^^^^^^^^^^^^^^^^^
       > AttributeError: 'AddrCacheTestCase' object has no attribute 'assertEquals'. Did you mean: 'assertEqual'?
       >
       > ----------------------------------------------------------------------
       > Ran 27 tests in 0.357s
       >
       > FAILED (errors=6, expected failures=1)
       > Test failed: <unittest.runner.TextTestResult run=27 errors=6 failures=0>
       > error: Test failed: <unittest.runner.TextTestResult run=27 errors=6 failures=0>
       For full logs, run 'nix log /nix/store/4b0yac479mdwca75ggwa4d276r1xhzwc-python3.12-pymilter-1.0.5.drv'.
error: 1 dependencies of derivation '/nix/store/x91p0xxgylixarbp6vdqdjqavg8fpqd3-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/308157

2 packages failed to build:
python312Packages.pymilter python312Packages.pymilter.dist

15 packages built:
exaile gramps gramps.dist python311Packages.bsddb3 python311Packages.bsddb3.dist python311Packages.pymilter python311Packages.pymilter.dist python311Packages.scrapy-deltafetch python311Packages.scrapy-deltafetch.dist python312Packages.bsddb3 python312Packages.bsddb3.dist python312Packages.scrapy-deltafetch python312Packages.scrapy-deltafetch.dist spf-engine spf-engine.dist
```

D3vil0p3r commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -1,55 +1,38 @@
-, fetchFromGitHub
-, python3
-}:
+{ lib, fetchFromGitHub, python3 }:
```
```suggestion
{ lib
, fetchFromGitHub
, python3
}:
```

AkechiShiro replied:
Thanks

D3vil0p3r reviewed (changes requested):
Not one-liner

AkechiShiro commented:
I think I fixed everything, the test was failing due to a deprecated setup pytest function, upstream has fixed the issue however `ssh-mitm` requires `Paramiko` <3.4 as such the test was failing when building with `Python312`, I introduced a patch to fix this issue.

@Aleksanaa hash was correctly updated too and format was upgraded to `pyproject`, this fixed the weird error happening with the hash.

I believe this PR is ready to be reviewed and hopefully merged

AkechiShiro commented:
Worth mentioning paramiko test failed due to this issue : https://github.com/paramiko/paramiko/issues/1959

AkechiShiro commented:
I'm not sure if backporting this PR is okay, I don't think there are any breaking change but this seems like a major release bump, maybe it shouldn't be backported to the latest stable release.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3922

AkechiShiro commented:
@D3vil0p3r any inputs on backporting this PR to stable release ?

AkechiShiro commented:
Anyone available for review if possible ? @fabaff maybe ?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1634

AkechiShiro commented:
Pretty weird, will attempt to fix it thanks for testing

AkechiShiro commented:
@marcusramberg can you test again, please ?

marcusramberg commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -1,55 +1,43 @@
     setuptools
     sshpubkeys
-  ];
+  ] ++ lib.optionals stdenv.darwin [ setuptools ];
```
This doesn't work, should probably be stdenv.isDarwin

AkechiShiro replied:
Yes sorry I just noticed the mistake, fixed it, now

AkechiShiro replied:
ofborg will probably evaluate too but I don't think it does for darwin, I may be wrong however.

AkechiShiro replied:
Good now ? @marcusramberg

fabaff commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -1,55 +1,43 @@
     sshpubkeys
-  ];
+  ] ++ lib.optionals stdenv.isDarwin [ setuptools ];
+  # fix for darwin users
+
+  #nativeBuildInputs = [ python3.pkgs.setuptools ];
```
```suggestion
```

AkechiShiro replied:
For sure, sorry for forgetting to add them, I'm still new to packaging well stuff in Nixpkgs, I will improve with your guidance.

fabaff commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -1,55 +1,43 @@
   format = "setuptools";
 
   src = fetchFromGitHub {
     owner = pname;
     repo = pname;
```
```suggestion
    owner = "ssh-mitm";
    repo = "ssh-mitm";
```

fabaff commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -1,55 +1,43 @@
   pname = "ssh-mitm";
-  version = "3.0.2";
+  version = "4.1.1";
   format = "setuptools";
```
```suggestion
  pyproject = true;
```

and then 

```
  build-system = with python3.pkgs; [
    hatchling
    hatch-requirements-txt
];
```

see https://github.com/ssh-mitm/ssh-mitm/blob/4.1.1/pyproject.toml#L1

fabaff commented on pkgs/tools/security/ssh-mitm/default.nix:
```diff
@@ -1,55 +1,43 @@
-{ lib
-, fetchFromGitHub
-, python3
+{ lib,
+fetchFromGitHub,
+python3,
+stdenv
 }:
```
```suggestion
{
  lib,
  stdenv,
  fetchFromGitHub,
  python3,
}:
```

to follow the latest RFC about formatting.

fabaff reviewed (commented):
There are manpages (https://github.com/ssh-mitm/ssh-mitm/tree/4.1.1/man1) could you please give it a shot to try to add them.

AkechiShiro commented:
Result of `nixpkgs-review pr 308093` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>ssh-mitm</li>
    <li>ssh-mitm.dist</li>
  </ul>
</details>

fabaff commented:
Result of `nixpkgs-review pr 308093` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>ssh-mitm</li>
    <li>ssh-mitm.dist</li>
  </ul>
</details>

fabaff reviewed (approved):
Change looks good to me.
